### PR TITLE
fix(AssetKitBlock): created date is not taken into account

### PR DESCRIPTION
**Pull request by fulopdaniel**
With the new approach we need to calculate the expiration timestamp with the following formula:

CreationDate + Lifetime.

Previously it was:

Date.now() + Lifetime

### CODE REVIEW

**Review by ragi96 — COMMENTED** [review]
But I think we should remove that cached url completely, and generate it every time freshly? wdyt?

File: packages/asset-kit-block/src/AssetKitBlock.tsx [hunk 1]
```diff
@@ -64,9 +64,22 @@ export const AssetKitBlock = ({ appBridge }: BlockProps): ReactElement => {
...
+    const transformDateStringToDate = (dateString: string) => {
+        const year = dateString.substring(0, 4);
+        const month = dateString.substring(4, 6);
+        const day = dateString.substring(6, 8);
+        const hour = dateString.substring(9, 11);
+        const minute = dateString.substring(11, 13);
+        const second = dateString.substring(13, 15);
+
+        return new Date(`${year}-${month}-${day}T${hour}:${minute}:${second}Z`);
+    };
```
[the remark this answers is not in the document]
Comment (fulopdaniel) [reply packages/asset-kit-block/src/AssetKitBlock.tsx]: I like the verbose version a bit better, it's more readable to me. But I did move the function to helpers to not occupy too much space in the main file.

File: packages/asset-kit-block/src/AssetKitBlock.tsx [hunk 2]
```diff
@@ -64,9 +64,22 @@ export const AssetKitBlock = ({ appBridge }: BlockProps): ReactElement => {
...
+
     const getExpirationTimestamp = (downloadUrl: string) => {
-        const expirationTimestamp = downloadUrl.split('X-Amz-Expires=')[1]?.split('&')[0] ?? 0;
-        return parseInt(expirationTimestamp, 10) + Math.floor(Date.now() / 1000);
+        const creationDate = transformDateStringToDate(downloadUrl.split('X-Amz-Date=')[1]?.split('&')[0] ?? '');
+        const lifetime = downloadUrl.split('X-Amz-Expires=')[1]?.split('&')[0] ?? 0;
+        const expirationTimestamp = creationDate.getTime() + Number(lifetime) * 1000;
+        return expirationTimestamp;
```
[the remark this answers is not in the document]
Comment (fulopdaniel) [reply packages/asset-kit-block/src/AssetKitBlock.tsx]: Fixed this one

**Comment by ragi96** [comment]
> days

**Comment by ragi96** [comment]
> > But I think we should remove that cached url completely, and generate it every time freshly? wdyt?
> 
> I like this caching approach, I think reducing network requests to one/click to one/5 days is pretty good, and I would keep it. We just need to get it right :)

I am also not really against it, but its bad as we are reading from urls we don't control. If something changes the caching will break and it won't be noticed

File: packages/asset-kit-block/src/helpers/assetKitHelpers.ts [hunk 3]
```diff
@@ -48,3 +48,14 @@ export const blockStyle = (blockSetting: Settings): CSSProperties => {
 export const getSmallPreviewUrl = (url: string) => {
     return url.replace('{width}', '218');
 };
+
+export const transformDateStringToDate = (dateString: string) => {
```
Comment (ragi96) [inline packages/asset-kit-block/src/helpers/assetKitHelpers.ts:52]: Can you add some tests for it

Comment (fulopdaniel) [reply packages/asset-kit-block/src/helpers/assetKitHelpers.ts:52]: Added

**Comment by fulopdaniel** [comment]
> > > But I think we should remove that cached url completely, and generate it every time freshly? wdyt?
> > 
> > 
> > I like this caching approach, I think reducing network requests to one/click to one/5 days is pretty good, and I would keep it. We just need to get it right :)
> 
> I am also not really against it, but its bad as we are reading from urls we don't control. If something changes the caching will break and it won't be noticed

Added an extra check to see if we successfully managed to parse the URL. If not, we generate the presigned URLs. wdyt?

**Review by ragi96 — APPROVED** [review]
LGTM

**Pull request merged**